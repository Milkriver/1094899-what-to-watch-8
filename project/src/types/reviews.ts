export interface IReview {
    id: number,
    movieId: number,
    author: string,
    date: string,
    rating: number,
    text: string,
  }
