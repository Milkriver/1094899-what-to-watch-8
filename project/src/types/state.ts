import { AuthorizationStatus } from '../const';
import { IMovie } from './common';
import { IReviewResponse } from './reviews';

export interface IState {
  currentGenre: string
  movies: IMovie[]
  isDataLoaded: boolean
  authorizationStatus: AuthorizationStatus,
  movie: IMovie,
  sameMovies: IMovie[],
  reviews: IReviewResponse[],
  promoMovie: IMovie,
}
