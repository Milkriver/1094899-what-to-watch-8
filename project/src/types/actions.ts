import { AxiosInstance } from 'axios';
import { Action } from 'redux';
import { ThunkAction, ThunkDispatch } from 'redux-thunk';
import { IState } from './state';

export enum ActionType {
  ChangeGenre = 'movieList/ChangeGenre',
  LoadMovies = 'movieList/LoadMovies',
  LoadSingleMovie = 'movieCard/LoadSingleMovie',
  LoadActiveMovie = 'movieCard/LoadActiveMovie',
  LoadSameGenreMovies = 'movieCard/LoadSameGenreMovies',
  LoadPromoMovie = 'movieCard/LoadPromoMovie',
  LoadFavouriteMovies = 'userList/LoadFavouriteMovies',
  AddFavoriteMovie = 'userList/AddFavoriteMovie',
  RemoveFavoriteMovie = 'userList/RemoveFavoriteMovie',
  LoadReviews = 'movieCard/LoadReviews',
  AddReview = 'movieCard/AddReview',
  RequireAuthorization = 'user/requireAuthorization',
  RequireLogout = 'user/requireLogout',
  LoadUserData = 'user/loadUserData',
}

export type ThunkActionResult<R = Promise<void>> = ThunkAction<R, IState, AxiosInstance, Action>;

export type ThunkAppDispatch = ThunkDispatch<IState, AxiosInstance, Action>;
