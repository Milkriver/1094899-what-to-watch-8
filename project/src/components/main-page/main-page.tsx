import { useState } from 'react';
import { useHistory } from 'react-router';
import { Link } from 'react-router-dom';
import { AppRoute } from '../../const';
import { IMovie } from '../../types/common';
import GenresList from '../genres-list/genres-list';
import MoviesList from '../movies-list/movies-list';
import { ShowMoreButton } from '../show-more-button/show-more-button';


type IProps = {
  cards: IMovie[]
}

export function MainPage({ cards }: IProps): JSX.Element {
  const history = useHistory();

  const [moviesShowingLimit, changeMoviesShowingLimit] = useState<number>(8);

  const [limit, changeLimit] = useState<number>(cards.length);
  const handleShowMoreButton = () => {
    const limitCondition = moviesShowingLimit + 8 < limit ? moviesShowingLimit + 8 : limit;
    changeMoviesShowingLimit(limitCondition);
  };
  const isButtonShowen = (moviesShowingLimit < limit);


  return (
    <div>

      <section className="film-card">
        <div className="film-card__bg">
          <img src={cards[0].background_image} alt="The Grand Budapest Hotel" />
        </div>

        <h1 className="visually-hidden">WTW</h1>

        <header className="page-header film-card__head">
          <div className="logo">
            <Link to={'/'} className="logo__link">
              <span className="logo__letter logo__letter--1">W</span>
              <span className="logo__letter logo__letter--2">T</span>
              <span className="logo__letter logo__letter--3">W</span>
            </Link>
          </div>

          <ul className="user-block">
            <li className="user-block__item">
              <div className="user-block__avatar">
                <img src="img/avatar.jpg" alt="User avatar" width="63" height="63" />
              </div>
            </li>
            <li className="user-block__item">
              <a href="/#" className="user-block__link">Sign out</a>
            </li>
          </ul>
        </header>

        <div className="film-card__wrap">
          <div className="film-card__info">
            <div className="film-card__poster">
              <img src="img/the-grand-budapest-hotel-poster.jpg" alt="The Grand Budapest Hotel poster" width="218" height="327" />
            </div>

            <div className="film-card__desc">
              <h2 className="film-card__title">{cards[0].name}</h2>
              <p className="film-card__meta">
                <span className="film-card__genre">{cards[0].genre}</span>
                <span className="film-card__year">{cards[0].released}</span>
              </p>

              <div className="film-card__buttons">
                <button className="btn btn--play film-card__button" type="button" onClick={() => history.push(AppRoute.Player)}>
                  <svg viewBox="0 0 19 19" width="19" height="19">
                    <use xlinkHref="#play-s"></use>
                  </svg>
                  <span>Play</span>
                </button>
                <button className="btn btn--list film-card__button" type="button" onClick={() => history.push(AppRoute.MyList)}>
                  <svg viewBox="0 0 19 20" width="19" height="20">
                    <use xlinkHref="#add"></use>
                  </svg>
                  <span>My list</span>
                </button>
              </div>
            </div>
          </div>
        </div>
      </section>

      <div className="page-content">
        <section className="catalog">
          <h2 className="catalog__title visually-hidden">Catalog</h2>
          <GenresList
            movies={cards}
          />

          <div className="catalog__films-list">
            <MoviesList
              changeLimit={changeLimit}
              cards={cards}
              moviesShowingLimit={moviesShowingLimit}
            />
          </div>
          <ShowMoreButton
            handleShowMoreButton={handleShowMoreButton}
            isButtonShowen={isButtonShowen}
          />
        </section>

        <footer className="page-footer">
          <div className="logo">
            <Link to={'/'} className="logo__link logo__link--light">
              <span className="logo__letter logo__letter--1">W</span>
              <span className="logo__letter logo__letter--2">T</span>
              <span className="logo__letter logo__letter--3">W</span>
            </Link>
          </div>

          <div className="copyright">
            <p>Â© 2019 What to watch Ltd.</p>
          </div>
        </footer>
      </div>
    </div>
  );
}
