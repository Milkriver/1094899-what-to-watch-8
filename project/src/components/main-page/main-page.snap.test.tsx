import { render } from '@testing-library/react';
import { createMemoryHistory } from 'history';
import { Router } from 'react-router';
import { Provider } from 'react-redux';
import { MainPage } from './main-page';
import { IMovie } from '../../types/common';
import { configureMockStore } from '@jedmao/redux-mock-store';

const mockMovie1: IMovie = {
  id: 1, name: 'name1', posterImage: 'posterImage1', previewImage: 'previewImage1', backgroundImage: 'backgroundImage1',
  backgroundColor: 'black', videoLink: 'videoLink1', previewVideoLink: 'previewVideoLink1',
  description: 'description1', rating: 1, scoresCount: 1, director: 'director1',
  starring: ['actor1', 'actor2'], runTime: 1, genre: 'Comedy', released: 1, isFavorite: true,
};

const mockMovie2: IMovie = {
  id: 2, name: 'name2', posterImage: 'posterImage2', previewImage: 'previewImage2', backgroundImage: 'backgroundImage2',
  backgroundColor: 'black', videoLink: 'videoLink2', previewVideoLink: 'previewVideoLink2',
  description: 'description2', rating: 2, scoresCount: 2, director: 'director2',
  starring: ['actor1', 'actor2'], runTime: 2, genre: 'Comedy', released: 2, isFavorite: true,
};

const mockMovies: IMovie[] = [mockMovie2, mockMovie1];
const mockFn = () => Promise.resolve();

describe('Component: MainPage', () => {
  const mockStore = configureMockStore();

  const store = mockStore({
    MOVIES: {},
    USER: {},
    REVIEWS: {},
  });

  store.dispatch = jest.fn();

  test('should render correctly', () => {
    const history = createMemoryHistory();
    const page = (
      <Provider store={store}>
        <Router history={history}>
          <MainPage cards={mockMovies} promoMovie={mockMovie1} onFetchPromoMovie={mockFn} onFetchActiveMovie={mockFn} />;
        </Router>
      </Provider>
    );
    const { container } = render(page);
    expect(container).toMatchSnapshot();
  });
});
