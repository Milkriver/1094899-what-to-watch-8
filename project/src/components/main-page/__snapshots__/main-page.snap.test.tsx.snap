// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: MainPage should render correctly 1`] = `
<div>
  <div>
    <section
      class="film-card"
    >
      <div
        class="film-card__bg"
      >
        <img
          alt="name1"
          src="backgroundImage1"
        />
      </div>
      <h1
        class="visually-hidden"
      >
        WTW
      </h1>
      <header
        class="page-header film-card__head"
      >
        <div
          class="logo"
        >
          <a
            class="logo__link"
            href="/"
          >
            <span
              class="logo__letter logo__letter--1"
            >
              W
            </span>
            <span
              class="logo__letter logo__letter--2"
            >
              T
            </span>
            <span
              class="logo__letter logo__letter--3"
            >
              W
            </span>
          </a>
        </div>
        <div
          class="user-block"
        >
          <a
            class="user-block__link"
            href="/login"
          >
            Sign in
          </a>
        </div>
      </header>
      <div
        class="film-card__wrap"
      >
        <div
          class="film-card__info"
        >
          <div
            class="film-card__poster"
          >
            <img
              alt="name1"
              height="327"
              src="posterImage1"
              width="218"
            />
          </div>
          <div
            class="film-card__desc"
          >
            <h2
              class="film-card__title"
            >
              name1
            </h2>
            <p
              class="film-card__meta"
            >
              <span
                class="film-card__genre"
              >
                Comedy
              </span>
              <span
                class="film-card__year"
              >
                1
              </span>
            </p>
            <div
              class="film-card__buttons"
            >
              <button
                class="btn btn--play film-card__button"
                type="button"
              >
                <svg
                  height="19"
                  viewBox="0 0 19 19"
                  width="19"
                >
                  <use
                    xlink:href="#play-s"
                  />
                </svg>
                <span>
                  Play
                </span>
              </button>
              <button
                class="btn btn--list film-card__button"
                type="button"
              >
                <svg
                  height="20"
                  viewBox="0 0 19 20"
                  width="19"
                >
                  <use
                    xlink:href="#in-list"
                  />
                </svg>
                <span>
                  My list
                </span>
              </button>
            </div>
          </div>
        </div>
      </div>
    </section>
    <div
      class="page-content"
    >
      <section
        class="catalog"
      >
        <h2
          class="catalog__title visually-hidden"
        >
          Catalog
        </h2>
        <ul
          class="catalog__genres-list"
        >
          <li
            class="catalog__genres-item"
          >
            <div
              class="catalog__genres-link "
            >
              All genres
            </div>
          </li>
          <li
            class="catalog__genres-item"
          >
            <div
              class="catalog__genres-link "
            >
              Comedy
            </div>
          </li>
        </ul>
        <div
          class="catalog__films-list"
        />
        <div
          class="catalog__more"
        />
      </section>
      <footer
        class="page-footer"
      >
        <div
          class="logo"
        >
          <a
            class="logo__link logo__link--light"
            href="/"
          >
            <span
              class="logo__letter logo__letter--1"
            >
              W
            </span>
            <span
              class="logo__letter logo__letter--2"
            >
              T
            </span>
            <span
              class="logo__letter logo__letter--3"
            >
              W
            </span>
          </a>
        </div>
        <div
          class="copyright"
        >
          <p>
            Â© 2019 What to watch Ltd.
          </p>
        </div>
      </footer>
    </div>
  </div>
  ;
</div>
`;
